### PR TITLE
Multiple version documentation

### PR DESCRIPTION
**Pull request by davfsa**
### Summary
This is a 2 part pull request. One of them is this one and the other is creating a repository like https://github.com/davfsa/hikari-docs.

Example render: https://davfsa.github.io/hikari-docs/

### Checklist

- [x] I have run `nox` and all the pipelines have passed.
- [x] I have made unittests according to the code I have added/modified/deleted.

### Related issues
#410

### CODE REVIEW

**Comment by davfsa** [comment]
Replace `hikari-py/hikari/.idea/fileTemplates/Mako Template.mako` with `hikari-py/hikari/.idea/fileTemplates/Jinja2 Template.jinja2`

**Comment by davfsa** [comment]
Think this is finally done.

A full render can be found here: https://davfsa.github.io/hikari-docs/stable/

*If anybody has a look and finds a bug/wants to improve something, feel free to leave a comment here, that repo or contact me on Discord*

File: docs/patched_pdoc.py
Comment (FasterSpeeding) [inline docs/patched_pdoc.py:98]: question, does this change fix/change the current behaviour where aliases for types/objects (e.g. `hikari.User` for `hikari.users.User`) aren't included in the generated objects.inv?

File: hikari/internal/ux.py
Comment (davfsa) [reply hikari/internal/ux.py:376]: > Does this actually change the behaviour at all or is this just a surplus style change?

It uses less memory and is a bit more efficient, since we only ever store what the version we are going to print is instead of storing all possible versions in a list and then sending the newest one

> Also you need to check for yanked releases

Yanked releases are already checked bellow.

File: hikari/internal/ux.py
Comment (davfsa) [reply hikari/internal/ux.py:324]: Will change

File: hikari/api/rest.py
Comment (davfsa) [reply hikari/api/rest.py:5288]: Wrong link, will fix

File: hikari/internal/deprecation.py
Comment (davfsa) [reply hikari/internal/deprecation.py:88]: When we follow a proper release schedule with proper versioning, this shouldnt be a problem, as the versions they are removed in will be mayor ones.

For now, where we only do dev releases, it will cause the CI to fail, but then we can quickly fix it.

-----

We really might want to have proper release schedule and figure out whether we want to have supported versions that get important fixes, etc.

(Ik you have been talking about this for a long time now, its just I dont have any concrete idea on how to do it properly. I am fully open to ideas an opinions)

File: docs/patched_pdoc.py
Comment (davfsa) [reply docs/patched_pdoc.py:98]: Yes. `hikari.User` will no longer be a thing in the `object.inv`. I considered it a bug and fixed it as if it were

File: hikari/internal/deprecation.py
Comment (FasterSpeeding) [reply hikari/internal/deprecation.py:88]: > For now, where we only do dev releases, it will cause the CI to fail, but then we can quickly fix it.

This would still be kind of annoying though, since it essentially means we both have to be active for a version to be released so we can deal with reviewing and merging any necessary follow up prs

File: hikari/api/rest.py
Comment (FasterSpeeding) [reply hikari/api/rest.py:6589]: Dead horse, don't really care anymore

File: hikari/impl/rest.py
Comment (FasterSpeeding) [reply hikari/impl/rest.py:3104]: don't care

File: hikari/api/rest.py
Comment (FasterSpeeding) [reply hikari/api/rest.py:6589]: hmm fair but i don't think this should block this pr either way

File: pipelines/pdoc.nox.py
Comment (davfsa) [reply pipelines/pdoc.nox.py:31]: `pages.nox.py` should have been removed, didnt realized it was still there. Will remove :+1:

File: hikari/internal/deprecation.py
Comment (FasterSpeeding) [reply hikari/internal/deprecation.py:88]: I don't really have any good systems for handling strict version based deadlines for stuff like this tbh; for my projects I'd usually give stuff a time based deprecation (so say 6 months) while they're in an alpha/beta state or if they're using a time based versioning scheme and once its stable I wouldn't really need a system like this which blares at me that a deprecated thing is still around since during a major version bump I'm going to have to comb through to find stuff to remove anyways since you can't have a system like this for arguments and it might actually give a false sense of security (possibly leading to stuff being missed) now that I think about it.

**Comment by FasterSpeeding** [comment]
What's the current state of this PR? I looked through it and couldn't really find anything else to point out (other than the fact that this pr somehow breaks linting but that might be fixed by a rebase) and this versioning stuff feels out of scope so i'd rather not have that keeping this opened.

File: hikari/internal/deprecation.py
Comment (davfsa) [reply hikari/internal/deprecation.py:88]: I am open to changing it to anything. I just cant think of a nice way that will warn us whenever we need to remove it.

I have thought about leaving the tests that will test the deprecation and then have `warn_deprecated` raise the error if needed, but that seemed a bit too much as it would make all deprecated calls slower.

**Comment by davfsa** [comment]
> What's the current state of this PR? I looked through it and couldn't really find anything else to point out (other than the fact that this pr somehow breaks linting but that might be fixed by a rebase) and this versioning stuff feels out of scope so i'd rather not have that keeping this opened.

It is mostly done. Once this is merged, I will open another PR to merge the feature branch into master and then I will create the repos and migrate the current docs to the new repo.

Might keep the feature branch opened for a day or two while I set everything up, as I dont have much time currently.

TLDR; Think it is good to go :)

**Comment by davfsa** [comment]
I just realized I have some TODO's left. Will try to get them done as soon as possible :sweat_smile:

File: hikari/internal/deprecation.py
Comment (FasterSpeeding) [reply hikari/internal/deprecation.py:88]: >  would make all deprecated calls slower.

Don't really see why that'd matter

> I am open to changing it to anything. I just cant think of a nice way that will warn us whenever we need to remove it.

The current system you've got in this PR doesn't achieve this since it only works for deprecated methods and not deprecated parameters

Comment (davfsa) [reply hikari/internal/deprecation.py:93]: > The current system you've got in this PR doesn't achieve this since it only works for deprecated methods and not deprecated parameters

Fixed that :+1:

Comment (FasterSpeeding) [reply hikari/internal/deprecation.py:93]: tw how'd you fix that?

**Review by FasterSpeeding — APPROVED** [review]

**Comment by davfsa** [comment]
This is finally done. The [render example](https://davfsa.github.io/hikari-docs/) shows how it would look and react once moved to https://docs.hikari-py.dev

Any feedback is welcome, but will be done as part of other PRs, as this is already getting too big and hard to rebase each time :P

**Pull request merged**
